File: typescript-playwright/src/clients/PhotoClient.ts
import type { PhotoApi } from "../api/PhotoApi";
import type { PhotoDto, PhotoPayload } from "../utils/types";
import { statusIn } from "../utils/statusIn";
import { HTTP } from "../utils/constants";
import { step } from "../utils/decorators";
export class PhotoClient {
  private readonly api: PhotoApi;

  constructor(api: PhotoApi) {
    this.api = api;
  }

  @step()
  async list(): Promise<PhotoDto[]> {
    const response = await this.api.getAll();
    if (!response.ok()) {
      throw new Error(
        `[PhotoClient] list failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<PhotoDto[]>;
  }

  @step()
  async listByCategory(categoryId: number): Promise<PhotoDto[]> {
    const response = await this.api.getByCategory(categoryId);
    if (!response.ok()) {
      throw new Error(
        `[PhotoClient] listByCategory(${categoryId}) failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<PhotoDto[]>;
  }

  @step()
  async getById(id: number): Promise<PhotoDto> {
    const response = await this.api.getById(id);
    if (!statusIn(HTTP.OK)(response.status())) {
      throw new Error(
        `[PhotoClient] getById(${id}) failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<PhotoDto>;
  }

  @step()
  async upload(filePath: string, payload: PhotoPayload): Promise<PhotoDto> {
    const response = await this.api.upload(filePath, payload);
    if (!statusIn(HTTP.OK, HTTP.CREATED)(response.status())) {
      throw new Error(
        `[PhotoClient] upload failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<PhotoDto>;
  }

  @step()
  async like(
    id: number,
  ): Promise<{ totalLikes: number; newlyLiked: boolean; message?: string }> {
    const response = await this.api.like(id);
    if (!statusIn(HTTP.OK)(response.status())) {
      throw new Error(
        `[PhotoClient] like(${id}) failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<{
      totalLikes: number;
      newlyLiked: boolean;
      message?: string;
    }>;
  }

  @step()
  async delete(id: number): Promise<void> {
    const response = await this.api.deletePhoto(id);
    if (!statusIn(HTTP.NO_CONTENT, HTTP.OK)(response.status())) {
      throw new Error(
        `[PhotoClient] delete(${id}) failed: ${response.status()} ${await response.text()}`,
      );
    }
  }

  @step()
  async tryDelete(id: number): Promise<void> {
    const response = await this.api.deletePhoto(id);
    if (
      !statusIn(HTTP.NO_CONTENT, HTTP.OK, HTTP.NOT_FOUND)(response.status())
    ) {
      throw new Error(
        `[PhotoClient] tryDelete(${id}) failed: ${response.status()} ${await response.text()}`,
      );
    }
  }
}

File: python/src/clients/photo_client.py
from __future__ import annotations

import allure
from pydantic import TypeAdapter

from src.api.photo_api import PhotoApi
from src.utils.constants import HTTP
from src.utils.models import LikeResult, PhotoDto, PhotoPayload

_PHOTO_LIST_ADAPTER = TypeAdapter(list[PhotoDto])


class PhotoClient:
    def __init__(self, api: PhotoApi) -> None:
        self._api = api

    @allure.step("API: Получить список всех фотографий")
    def photo_list(self) -> list[PhotoDto]:
        response = self._api.get_all()
        if not response.ok:
            raise RuntimeError(f"[PhotoClient] list failed: {response.status_code} {response.text}")
        return _PHOTO_LIST_ADAPTER.validate_python(response.json())

    @allure.step("API: Получить фотографии из категории ID {category_id}")
    def list_by_category(self, category_id: int) -> list[PhotoDto]:
        response = self._api.get_by_category(category_id)
        if not response.ok:
            raise RuntimeError(
                f"[PhotoClient] listByCategory({category_id}) failed: {response.status_code} {response.text}"
            )
        return _PHOTO_LIST_ADAPTER.validate_python(response.json())

    @allure.step("API: Получить фотографию по ID {photo_id}")
    def get_by_id(self, photo_id: int) -> PhotoDto:
        response = self._api.get_by_id(photo_id)
        if response.status_code not in (HTTP.OK,):
            raise RuntimeError(
                f"[PhotoClient] getById({photo_id}) failed: {response.status_code} {response.text}"
            )
        return PhotoDto.model_validate(response.json())

    @allure.step("API: Загрузить файл из '{file_path}'")
    def upload(self, file_path: str, payload: PhotoPayload) -> PhotoDto:
        response = self._api.upload(file_path, payload)
        if response.status_code not in (HTTP.OK, HTTP.CREATED):
            raise RuntimeError(
                f"[PhotoClient] upload failed: {response.status_code} {response.text}"
            )
        return PhotoDto.model_validate(response.json())

    @allure.step("API: Поставить лайк фотографии ID {photo_id}")
    def like(self, photo_id: int) -> LikeResult:
        response = self._api.like(photo_id)
        if response.status_code not in (HTTP.OK,):
            raise RuntimeError(
                f"[PhotoClient] like({photo_id}) failed: {response.status_code} {response.text}"
            )
        return LikeResult.model_validate(response.json())

    @allure.step("API: Обязательное удаление фотографии ID {photo_id}")
    def delete(self, photo_id: int) -> None:
        response = self._api.delete_photo(photo_id)
        if response.status_code not in (HTTP.NO_CONTENT, HTTP.OK):
            raise RuntimeError(
                f"[PhotoClient] delete({photo_id}) failed: {response.status_code} {response.text}"
            )

    @allure.step("API: Попытка удаления фотографии ID {photo_id} (игнорируя 404)")
    def try_delete(self, photo_id: int) -> None:
        response = self._api.delete_photo(photo_id)
        if response.status_code not in (HTTP.NO_CONTENT, HTTP.OK, HTTP.NOT_FOUND):
            raise RuntimeError(
                f"[PhotoClient] tryDelete({photo_id}) failed: {response.status_code} {response.text}"
            )
